### DISCUSSION TITLE
OpenID Connect support

### DISCUSSION THREAD

**Discussion by utkarsh-bandekar** [invertase/react-native-firebase#5957] [General]
We have a react-native app which is currently using Auth0 for authentication.

We are planning to add SSO support in our app. We have decided to use Google identity platform and configure OpenID Connect provider for tenants which requires SSO.

Does react-native-firebase support login using OpenID Connect provider.

If its not supported then how much easy it will be to add support for OpenID Connect(Considering the fact that we are new in react-native development)

**Comment by paulrostorp** [comment]
I am also interested in this, is this something feasible ?

**Reply by mikehardy** [reply]
Hey @paulrostorp :wave:  (and also @utkarsh-bandekar - sorry about the long long delay in answering here)

This is similar to #5401 and #6305 (both focused on adding Microsoft as an auth provider, but the theme is the same: it's a feature gap here right now)

The general idea for "I want to get this done without blocking on anyone" is to use https://github.com/FormidableLabs/react-native-app-auth/ to get your identity/auth token from identity provider, then use firebase-admin-sdk via firebase cloud functions to mint a firebase authentication token so you may authorize firebase services

That will work but is project-specific so in a sense represents future "wasted" effort as it will require project-specific maintenance vs leveraging the community and maintenance work here

The more general solution would be to add support for the generic oauth provider firebase has, so we can expand this list (https://github.com/FormidableLabs/react-native-app-auth/) to contain "generic OAuth provider" :-)

https://firebase.google.com/docs/reference/android/com/google/firebase/auth/OAuthProvider
https://firebase.google.com/docs/reference/ios/firebaseauth/api/reference/Classes/FIROAuthProvider

And a general template for how to add a new provider is probably our latest addition a while back, when Sign On With Apple was added: 
https://github.com/invertase/react-native-firebase/pull/2979/files

The generic OAuth provider has been used once already for android platform apple signon - seems easy to work with:
https://github.com/invertase/react-native-firebase/pull/4188/files

If a PR came in here for OAuth support then it gets maintained forever by everyone and you wouldn't need the firebase-admin-sdk/cloud-functions intermediary, as react-native-firebase would use the underlying native SDKs APIs to mint the firebase token from your OAuth token

**Reply by paulrostorp** [reply]
@mikehardy I actually spent a couple of hours today working with the native code trying to get it to work with the generic firebase oAuth provider. I tried with both IOS and android without success, but perhaps my approach was wrong.
I first setup https://www.npmjs.com/package/react-native-azure-auth and succesfully got an id token and access token.
I then passed those to a modified firebase `signInWithCredentials`. I tried with both firebase oauth provider "microsoft.com" and "oidc.microsoft" (https://cloud.google.com/identity-platform/docs/web/oidc).
I kept getting an error, something about a "malformed token".

Any advice ?

**Reply by mikehardy** [reply]
Hmm I'm reading these closely now (I had not before, apologies) https://firebase.google.com/docs/auth/android/microsoft-oauth / https://firebase.google.com/docs/auth/ios/microsoft-oauth#objective-c - it appears that microsoft is supported internal to the firebase SDKs so the react-native-azure-auth + get token step should not be required, it appears to be more a matter of configuring the Azure secrets in the Firebase console for your project (perhaps downloading new plist/json config files) then "simply" creating the microsoft.com OAuthProvider and calling the underlying getCredientialWithUIDelegate / startActivityForSignInWithProvider methods

At that point, the native SDKs should pop a web Custom Tab to do the auth via microsoft and handle the microsoft token -> firebase token stuff, returning it all as a nice auth result with OAuthCredential inside ?
